File: problem-4.js

//returns true if the array contains elements in a palindome order
var isPalindrome = function(a){
	var palindrome = true;
	for(var i = 0; i < a.length; i++){
		if(a[i] != a[a.length - 1 - i]){
			palindrome = false;
		};
	};
	return palindrome;
};

//returns an array of each digit of a number
var numToArray = function(num){
	var a = [];
	while(num > 0){
		a.push(num%10);
		num = Math.floor(num/10);
	}
	return a;
};

//returns an array of the products of numbers with the given number of digits.
//i.e. The products of numbers 1-9 (for digit = 1)
var products = function(digit){
	var products = [];
	for(var i = Math.pow(10,digit)/10; i<Math.pow(10,digit+1)/10; i++){
		for(var j = Math.pow(10,digit)/10; j<Math.pow(10,digit+1)/10; j++){
			products.push(i*j);
		};
	};
	return products;
};

//returns an array of palindromes
//takes an array of numbers
var palindromeArray = function(nums){
	var a = [];
	for(var i = 0; i < nums.length; i++){
		if(isPalindrome(numToArray(nums[i]))){
			a.push(nums[i]);
		};
	};
	return a;
};

//returns the largest value in an array
var findLargest = function(a){
	var big = 0;
	for(var i = 0; i < a.length; i++){
		if(a[i] > big){
			big = a[i];
		};
	};
	return big;
};
print(numToArray(10901));
print(isPalindrome(10901));

print(findLargest(palindromeArray(products(3))));

File: problem-4.rb
#A palindromic number reads the same both ways.
#The largest palindrome made from the product of
#two 2-digit numbers is 9009 = 91 × 99.

#Find the largest palindrome made from the product of two 3-digit numbers.

#pseudo-code
#generate factors
#iterate backwards through factors until a palindrome is found

#generates all the factors of 2 numbers of the given digit size
#returns the factors in an array
def generate_factors(digit)
	factors = []
	for i in 10**digit/10...10**digit
		for j in i...10**digit
			factors<<j*i
		end
	end
	factors
end

#returns true if the given number is a palindrome
def is_palindrome?(number)
	number.to_s.reverse == number.to_s
end

def find_largest_palindrome(factors)
	largest = 0
	factors.reverse.each do |factor|
		largest = factor
		break if is_palindrome?(largest)
	end
	largest
end

puts find_largest_palindrome generate_factors(3).uniq.sort